### ISSUE TITLE
Custom trained model but no idea how to improve it.

### ISSUE THREAD

**Issue by shiennie** [matterport/Mask_RCNN#2547]
Hi all,

I am new to Mask RCNN. I trained a small dataset with 7 classes using 210 images for training and 28 for validation. This is my configuration. Any suggestion on how to improve it ?

>Configurations:
BACKBONE                       resnet50
BACKBONE_STRIDES               [4, 8, 16, 32, 64]
BATCH_SIZE                     1
BBOX_STD_DEV                   [0.1 0.1 0.2 0.2]
COMPUTE_BACKBONE_SHAPE         None
DETECTION_MAX_INSTANCES        400
DETECTION_MIN_CONFIDENCE       0.5
DETECTION_NMS_THRESHOLD        0.3
FPN_CLASSIF_FC_LAYERS_SIZE     1024
GPU_COUNT                      1
GRADIENT_CLIP_NORM             5.0
IMAGES_PER_GPU                 1
IMAGE_CHANNEL_COUNT            3
IMAGE_MAX_DIM                  512
IMAGE_META_SIZE                20
IMAGE_MIN_DIM                  512
IMAGE_MIN_SCALE                2.0
IMAGE_RESIZE_MODE              crop
IMAGE_SHAPE                    [512 512   3]
LEARNING_MOMENTUM              0.9
LEARNING_RATE                  0.001
LOSS_WEIGHTS                   {'rpn_class_loss': 1.0, 'rpn_bbox_loss': 1.0, 'mrcnn_class_loss': 1.0, 'mrcnn_bbox_loss': 1.0, 'mrcnn_mask_loss': 1.0}
MASK_POOL_SIZE                 14
MASK_SHAPE                     [28, 28]
MAX_GT_INSTANCES               200
MEAN_PIXEL                     [43.53 39.56 48.22]
MINI_MASK_SHAPE                (56, 56)
NAME                           custom
NUM_CLASSES                    8
POOL_SIZE                      7
POST_NMS_ROIS_INFERENCE        2000
POST_NMS_ROIS_TRAINING         1000
PRE_NMS_LIMIT                  6000
ROI_POSITIVE_RATIO             0.33
RPN_ANCHOR_RATIOS              [0.5, 1, 2]
RPN_ANCHOR_SCALES              (16, 32, 64, 128, 256)
RPN_ANCHOR_STRIDE              1
RPN_BBOX_STD_DEV               [0.1 0.1 0.2 0.2]
RPN_NMS_THRESHOLD              0.9
RPN_TRAIN_ANCHORS_PER_IMAGE    64
STEPS_PER_EPOCH                100
TOP_DOWN_PYRAMID_SIZE          256
TRAIN_BN                       False
TRAIN_ROIS_PER_IMAGE           64
USE_MINI_MASK                  False
USE_RPN_ROIS                   True
VALIDATION_STEPS               25
WEIGHT_DECAY                   0.0001

![graph](https://user-images.githubusercontent.com/17925020/116336755-b170d300-a80b-11eb-9832-0841435eda15.PNG)

**Comment by rakesh-kumar-k** [comment]
> Hi all,
> 
> I am new to Mask RCNN. I trained a small dataset with 7 classes using 210 images for training and 28 for validation. This is my configuration. Any suggestion on how to improve it ?
> 
> > Configurations:
> > BACKBONE                       resnet50
> > BACKBONE_STRIDES               [4, 8, 16, 32, 64]
> > BATCH_SIZE                     1
> > BBOX_STD_DEV                   [0.1 0.1 0.2 0.2]
> > COMPUTE_BACKBONE_SHAPE         None
> > DETECTION_MAX_INSTANCES        400
> > DETECTION_MIN_CONFIDENCE       0.5
> > DETECTION_NMS_THRESHOLD        0.3
> > FPN_CLASSIF_FC_LAYERS_SIZE     1024
> > GPU_COUNT                      1
> > GRADIENT_CLIP_NORM             5.0
> > IMAGES_PER_GPU                 1
> > IMAGE_CHANNEL_COUNT            3
> > IMAGE_MAX_DIM                  512
> > IMAGE_META_SIZE                20
> > IMAGE_MIN_DIM                  512
> > IMAGE_MIN_SCALE                2.0
> > IMAGE_RESIZE_MODE              crop
> > IMAGE_SHAPE                    [512 512   3]
> > LEARNING_MOMENTUM              0.9
> > LEARNING_RATE                  0.001
> > LOSS_WEIGHTS                   {'rpn_class_loss': 1.0, 'rpn_bbox_loss': 1.0, 'mrcnn_class_loss': 1.0, 'mrcnn_bbox_loss': 1.0, 'mrcnn_mask_loss': 1.0}
> > MASK_POOL_SIZE                 14
> > MASK_SHAPE                     [28, 28]
> > MAX_GT_INSTANCES               200
> > MEAN_PIXEL                     [43.53 39.56 48.22]
> > MINI_MASK_SHAPE                (56, 56)
> > NAME                           custom
> > NUM_CLASSES                    8
> > POOL_SIZE                      7
> > POST_NMS_ROIS_INFERENCE        2000
> > POST_NMS_ROIS_TRAINING         1000
> > PRE_NMS_LIMIT                  6000
> > ROI_POSITIVE_RATIO             0.33
> > RPN_ANCHOR_RATIOS              [0.5, 1, 2]
> > RPN_ANCHOR_SCALES              (16, 32, 64, 128, 256)
> > RPN_ANCHOR_STRIDE              1
> > RPN_BBOX_STD_DEV               [0.1 0.1 0.2 0.2]
> > RPN_NMS_THRESHOLD              0.9
> > RPN_TRAIN_ANCHORS_PER_IMAGE    64
> > STEPS_PER_EPOCH                100
> > TOP_DOWN_PYRAMID_SIZE          256
> > TRAIN_BN                       False
> > TRAIN_ROIS_PER_IMAGE           64
> > USE_MINI_MASK                  False
> > USE_RPN_ROIS                   True
> > VALIDATION_STEPS               25
> > WEIGHT_DECAY                   0.0001
> 
> ![graph](https://user-images.githubusercontent.com/17925020/116336755-b170d300-a80b-11eb-9832-0841435eda15.PNG)
 


Will you please tell how to train on custom dataset? As I am finding errors on trying on my own!

**Comment by shiennie** [comment]
@Rakeshkangal I think you can refer this post https://medium.com/analytics-vidhya/a-simple-guide-to-maskrcnn-custom-dataset-implementation-27f7eab381f2 . Hope this gives you some idea on how to train.

**Comment by shiennie** [comment]
#527

**Comment by ssocean** [comment]
Given that your class_loss seems unstable, I think you should improve the classification branch...

**Comment by shiennie** [comment]
@ssocean Can further explain on the improvement to classification branch ? Is it like tuning the parameters stated in the post ? For now, whenever I do prediction on cervical cell, there are still possibilities that the model predicted the correct class but different severity. For example, there are 3 severity in Dysplasia class for cervical cell, I could get the prediction of normal dysplasia, and severe dysplasia while the pictures I input is mild dysplasia.

**Comment by ssocean** [comment]
How do these cervical cells look like? Do they present great variation? If they looks simlar, I'afraid that tuning the parameters is not enough. 

I think either the feature selection capability of FPN or the classification capability  of the classification layey is insufficient. As can be found in model.py, fpn_classifier_graph intend to classify object's features from FPN by only one single Dense(Fully Connection) layer. 

Here're my suggestion: (1) increase the size of mask shape a little bigger.  As mentioned in https://github.com/matterport/Mask_RCNN/issues/635#issue-328851469 (2) Conduct an experiment of the cell classification with FPN + two conv layer + one dense layer. The input of the model is cropped patch containing the cervical cell from the whole image. Dataset can be small even there could be overfitting. If acc of training set is quite low, increase a few dense layers. Try again, if it is still low, try improve FPN. If failed again, stop using Mask R-CNN.

Hopefully this will help you!  Please let me know if you solve this problem.

![image](https://user-images.githubusercontent.com/48676611/117973515-16294300-b35f-11eb-92a5-b301cd6dce18.png)

**Comment by ssocean** [comment]
One more thing, Could you please explain why the class_loss climbs up around epoch 9 suddenly? Have you shuffled the data? Have you used techniques to change learning rate under certain situation? It doesn't look like the gradient explosion, more like data problem. This could be the key to your issue.

**Comment by shiennie** [comment]
> How do these cervical cells look like? Do they present great variation? If they looks simlar, I'afraid that tuning the parameters is not enough.
> 
> I think either the feature selection capability of FPN or the classification capability of the classification layey is insufficient. As can be found in model.py, fpn_classifier_graph intend to classify object's features from FPN by only one single Dense(Fully Connection) layer.
> 
> Here're my suggestion: (1) increase the size of mask shape a little bigger. As mentioned in [#635 (comment)](https://github.com/matterport/Mask_RCNN/issues/635#issue-328851469) (2) Conduct an experiment of the cell classification with FPN + two conv layer + one dense layer. The input of the model is cropped patch containing the cervical cell from the whole image. Dataset can be small even there could be overfitting. If acc of training set is quite low, increase a few dense layers. Try again, if it is still low, try improve FPN. If failed again, stop using Mask R-CNN.
> 
> Hopefully this will help you! Please let me know if you solve this problem.
> 
> ![image](https://user-images.githubusercontent.com/48676611/117973515-16294300-b35f-11eb-92a5-b301cd6dce18.png)

@ssocean Currently, I am using Herlev dataset for my training and validation sets. All the images were annotated using VIA. The image looks something like this. 

![cervical cell example](https://user-images.githubusercontent.com/17925020/118106161-e4bc8000-b40f-11eb-84b7-8f3c693b10d7.PNG)



> One more thing, Could you please explain why the class_loss climbs up around epoch 9 suddenly? Have you shuffled the data? Have you used techniques to change learning rate under certain situation? It doesn't look like the gradient explosion, more like data problem. This could be the key to your issue.

I think should be insufficient dataset, I only have 30 images each for 7 classes in training and 4 images for testing. And I using high step_per_epoch and validation steps, guess I over trained the model, causing overfitting.

**Comment by AlenaKV** [comment]
> Hi all,
> 
> I am new to Mask RCNN. I trained a small dataset with 7 classes using 210 images for training and 28 for validation. This is my configuration. Any suggestion on how to improve it ?
> 
> > Configurations:
> > BACKBONE                       resnet50
> > BACKBONE_STRIDES               [4, 8, 16, 32, 64]
> > BATCH_SIZE                     1
> > BBOX_STD_DEV                   [0.1 0.1 0.2 0.2]
> > COMPUTE_BACKBONE_SHAPE         None
> > DETECTION_MAX_INSTANCES        400
> > DETECTION_MIN_CONFIDENCE       0.5
> > DETECTION_NMS_THRESHOLD        0.3
> > FPN_CLASSIF_FC_LAYERS_SIZE     1024
> > GPU_COUNT                      1
> > GRADIENT_CLIP_NORM             5.0
> > IMAGES_PER_GPU                 1
> > IMAGE_CHANNEL_COUNT            3
> > IMAGE_MAX_DIM                  512
> > IMAGE_META_SIZE                20
> > IMAGE_MIN_DIM                  512
> > IMAGE_MIN_SCALE                2.0
> > IMAGE_RESIZE_MODE              crop
> > IMAGE_SHAPE                    [512 512   3]
> > LEARNING_MOMENTUM              0.9
> > LEARNING_RATE                  0.001
> > LOSS_WEIGHTS                   {'rpn_class_loss': 1.0, 'rpn_bbox_loss': 1.0, 'mrcnn_class_loss': 1.0, 'mrcnn_bbox_loss': 1.0, 'mrcnn_mask_loss': 1.0}
> > MASK_POOL_SIZE                 14
> > MASK_SHAPE                     [28, 28]
> > MAX_GT_INSTANCES               200
> > MEAN_PIXEL                     [43.53 39.56 48.22]
> > MINI_MASK_SHAPE                (56, 56)
> > NAME                           custom
> > NUM_CLASSES                    8
> > POOL_SIZE                      7
> > POST_NMS_ROIS_INFERENCE        2000
> > POST_NMS_ROIS_TRAINING         1000
> > PRE_NMS_LIMIT                  6000
> > ROI_POSITIVE_RATIO             0.33
> > RPN_ANCHOR_RATIOS              [0.5, 1, 2]
> > RPN_ANCHOR_SCALES              (16, 32, 64, 128, 256)
> > RPN_ANCHOR_STRIDE              1
> > RPN_BBOX_STD_DEV               [0.1 0.1 0.2 0.2]
> > RPN_NMS_THRESHOLD              0.9
> > RPN_TRAIN_ANCHORS_PER_IMAGE    64
> > STEPS_PER_EPOCH                100
> > TOP_DOWN_PYRAMID_SIZE          256
> > TRAIN_BN                       False
> > TRAIN_ROIS_PER_IMAGE           64
> > USE_MINI_MASK                  False
> > USE_RPN_ROIS                   True
> > VALIDATION_STEPS               25
> > WEIGHT_DECAY                   0.0001
> 
> ![graph](https://user-images.githubusercontent.com/17925020/116336755-b170d300-a80b-11eb-9832-0841435eda15.PNG)

Hey! I have a stupid question, but how did you can build these graphs? I work in Google Colab and don't understand what command to build these graphs.Thanks!

**Comment by ssocean** [comment]
These graphs were generated by Tensorboard, the log file looks like 'events.out.tfevents.***********.USERNAME'.

**Comment by shiennie** [comment]
@AlenaKV as @ssocean  said this is generated using tensorboard locally, there are some commands to call tensorboard using google colab, could reference the link https://colab.research.google.com/github/tensorflow/tensorboard/blob/master/docs/tensorboard_in_notebooks.ipynb#scrollTo=TB0wBWfcVqHz hope it helps you.
